### PR TITLE
Use lightningcss to implement CSS packager

### LINKED ISSUE #8162: CSS `@import` layers are ignored in some situations
# 🐛 bug report

CSS `@import` layers are ignored in some situations. For example:

```css
@import 'red.css' layer;

body {
  background: green;
}
```

## 🎛 Configuration (.babelrc, package.json, cli command)

```json
{
  "name": "parcel-css-layer-import-repro",
  "version": "1.0.0",
  "scripts": {
    "start": "parcel src/index.html"
  },
  "browserslist": "last 1 Chrome version",
  "dependencies": {
    "parcel": "2.6.0"
  }
}
```

## 🤔 Expected Behavior

The `body` should display green with css like this:

```css
@layer {
  body.body {
    background: red;
  }
}

body {
  background: green;
}
```

## 😯 Current Behavior

The `body` displays as red because the css is compiled without layers:

```css
body.body {
  background: red;
}

body {
  background: green;
}

```

## 🔦 Context

In my actual project, the non-layer override is defined dynamically via css-in-js. An example of this can be found in [the `react` branch of the repro repo](https://github.com/matt-tingen/parcel-css-layer-import-repro/tree/react). I mention this because I'm not sure if parcel(-css) is trying to use static analysis to avoid  `@layer` for compat, but doing so seems invalid when all of the declared browsers support `@layer`.

## 💻 Code Sample

[Repro repo](https://github.com/matt-tingen/parcel-css-layer-import-repro)

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | 2.6.0
| Node             | 16.13.0
| npm/Yarn         | yarn 1.22.17
| Operating System | macOS 12.4

### PR DESCRIPTION
**Pull request by devongovett**
Fixes #8162

This re-implements the CSS packager using the `bundleAsync` function of `lightningcss`. This means there is support for bundling `@import` rules with cascade layers and supports conditions, and improved support for media queries. The logic for bundling these is fairly complicated, and can't really be done fully correctly without an AST.

Need to do some testing on real world apps, and some perf testing.

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 1.43s | -41.00ms |
| Cached | 348.00ms | -3.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

<details><summary>React HackerNews 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

<details><summary>AtlasKit Editor 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

<details><summary>Three.js ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 6.89s | +46.00ms |
| Cached | 295.00ms | +4.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/343294124950028884)**

File: packages/packagers/css/src/CSSPackager.js [hunk 1]
```diff
@@ -106,77 +134,112 @@ export default (new Packager({
...
+      (await queue.run()).map(([asset, code, map]) => [asset, [code, map]]),
+    );
     let map = new SourceMap(options.projectRoot);
-    let lineOffset = 0;
 
-    for (let url of hoistedImports) {
-      contents += `@import "${url}";\n`;
-      lineOffset++;
-    }
+    let res = await bundleAsync({
```
Comment (mischnic) [inline packages/packagers/css/src/CSSPackager.js:157]: What can we do for the REPL where `bundleAsync` doesn't exist?

Comment (devongovett) [reply packages/packagers/css/src/CSSPackager.js:157]: It does exist now (as of a few months ago)! Uses emscripten's asyncify transform to turn the rust function into a coroutine.

Comment (mischnic) [reply packages/packagers/css/src/CSSPackager.js:157]: Ah, perfect!

Comment (mischnic) [reply packages/packagers/css/src/CSSPackager.js:157]: We will need to usual alias (and devdep) to the wasm package though like in the CSS packager and optimizer:
https://github.com/parcel-bundler/parcel/blob/6801029b7ad89fc264ff5554e95ee0fcd6fe699c/packages/optimizers/css/package.json#L34-L36

File: packages/packagers/css/src/CSSPackager.js [hunk 2]
```diff
@@ -106,77 +134,112 @@ export default (new Packager({
...
+          if (!asset) {
+            return '';
+          }
+          let [code, map] = nullthrows(outputs.get(asset));
+          if (map) {
+            let sm = await map.stringify({format: 'inline'});
+            invariant(typeof sm === 'string');
+            code += `\n/*# sourceMappingURL=${sm} */`;
+          }
+          return code;
```
Comment (mischnic) [inline packages/packagers/css/src/CSSPackager.js:179]: Isn't this repeated stringification, inlining and parsing of the sourcemap pretty slow?

**Review by mischnic — APPROVED** [review]

File: packages/packagers/css/src/CSSPackager.js (hunk 2 above)
Comment (devongovett) [reply packages/packagers/css/src/CSSPackager.js:179]: probably not ideal. once we move to rust we can just pass around a parcel source map instance. lightningcss uses the same library internally but the JS API still returns it as a JSON buffer.

File: packages/packagers/css/src/CSSPackager.js (hunk 1 above)
Comment (devongovett) [reply packages/packagers/css/src/CSSPackager.js:157]: added the dep. are we just relying on the transformer to initialize the wasm module? looks like CSSOptimizer doesn't do it, so I just followed that.

Comment (mischnic) [reply packages/packagers/css/src/CSSPackager.js:157]: Good point, I apparently forgot that in the optimizer. It's not really safe to assume that every worker running the packager also has also at least once invoked the CSS transformer.

Comment (devongovett) [reply packages/packagers/css/src/CSSPackager.js:157]: ok added to both

**Pull request merged**